lib/engine-universal/src/executable.rs

```diff
@@ -0,0 +1,286 @@
+    pub(crate) data_initializers: Vec<OwnedDataInitializer>,
+    pub(crate) cpu_features: u64,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ExecutableSerializeError {
+    #[error("could not serialize the executable data")]
+    Executable(
+        #[source]
+        CompositeSerializerError<
+            std::convert::Infallible,
+            AllocScratchError,
+            SharedSerializeMapError,
+        >,
+    ),
+}
+
+// SAFETY: the pointers in `rkyv::AllocScratchError` are present there for display purposes – this
+// type does not expose any mechanism to access or operate on the pointers, unless they are
+// accessed directly by the user.
+unsafe impl Send for ExecutableSerializeError {}
+unsafe impl Sync for ExecutableSerializeError {}
+
+impl wasmer_engine::Executable for UniversalExecutable {
+    fn load(
+        &self,
+        engine: &(dyn Engine + 'static),
+    ) -> Result<std::sync::Arc<dyn Artifact>, CompileError> {
+        engine
+            .downcast_ref::<crate::UniversalEngine>()
+            .ok_or_else(|| CompileError::Codegen("can't downcast TODO FIXME".into()))?
+            .load_universal_executable(self)
+            .map(|a| Arc::new(a) as _)
+    }
+
+    fn features(&self) -> Features {
+        self.compile_info.features.clone()
+    }
+
+    fn cpu_features(&self) -> EnumSet<CpuFeature> {
+        EnumSet::from_u64(self.cpu_features)
+    }
+
+    fn serialize(&self) -> Result<Vec<u8>, Box<(dyn std::error::Error + Send + Sync + 'static)>> {
+        // The format is as thus:
+        //
+        // HEADER
+        // RKYV PAYLOAD
+        // RKYV POSITION
+        //
+        // It is expected that any framing for message length is handled by the caller.
+        let mut out = Vec::with_capacity(32);
+        out.extend(&MAGIC_HEADER);
+        let mut serializer = AllocSerializer::<1024>::default();
+        let pos = rkyv::ser::Serializer::serialize_value(&mut serializer, self)
+            .map_err(ExecutableSerializeError::Executable)? as u64;
+        out.extend(serializer.into_serializer().into_inner().as_slice());
+        out.extend(&pos.to_le_bytes());
+        Ok(out)
+    }
+
+    fn function_name(&self, index: FunctionIndex) -> Option<&str> {
+        let module = &self.compile_info.module;
+        // First, lets see if there's a name by which this function is exported.
+        for (name, idx) in module.exports.iter() {
+            match idx {
+                &ExportIndex::Function(fi) if fi == index => return Some(&*name),
+                _ => continue,
+            }
+        }
+        if let Some(r) = module.function_names.get(&index) {
+            return Some(&**r);
+        }
+        for ((_, field, _), idx) in module.imports.iter() {
+            match idx {
+                &ImportIndex::Function(fi) if fi == index => return Some(&*field),
+                _ => continue,
+            }
+        }
+        None
+    }
+
+    fn make_local_function_index(
```

Comment:
TODO: change this to return an `EntityCounts` with the relevant methods implemented on it.